juju/juju
# raft: Enable raft workers in the machine agent

## Description of change

This runs the raft workers in the machine agent, as a step towards managing leases and leadership with raft. The new workers are:

raft-transport: manages connections between the raft nodes (runs in all controller agents).
raft: runs the raft node and provides it to other workers (all controllers).
raft-flag: bounces whenever the raft leadership of this node changes and exposes a .Check method that returns whether this is the leader (all controllers).
raft-clusterer: watches for changes to the set of controller machines and updates raft as needed (only runs on the leader node).

At the moment the payload the raft cluster manages is rudimentary - it just maintains a log of entries. In the future this will be replaced with an FSM that manages leases.

## QA steps

* Bootstrap a controller, deploy some applications and run `juju enable-ha`. 
* Watch the debug logging for juju.worker.raft (and modules under that) to see the controller machines get added to the cluster and elect a leader.
* SSH to the controller machines and use `juju-engine-report` to check the raft cluster is as expected.
* Upgrade juju and see the raft workers come back up and resume normal interaction.
* Stop jujud on the leader machine and see that the raft cluster elects a new leader.
* Restart jujud on the old leader and see that the raft workers catch up.
